hazelcast/src/main/java/com/hazelcast/internal/partition/impl/MigrationManager.java

```diff
@@ -737,27 +744,38 @@ private void assignCompletelyLostPartitions() {
 
             logger.fine("Cluster state doesn't allow repartitioning. RepartitioningTask will only assign lost partitions.");
             InternalPartition[] partitions = partitionStateManager.getPartitions();
-            PartitionIdSet partitionIds = new PartitionIdSet(partitions.length);
-            for (InternalPartition partition : partitions) {
-                boolean empty = true;
-                for (int index = 0; index < InternalPartition.MAX_REPLICA_COUNT; index++) {
-                    empty &= partition.getReplica(index) == null;
-                }
-                if (empty) {
-                    partitionIds.add(partition.getPartitionId());
-                }
-            }
+            PartitionIdSet partitionIds = Arrays.stream(partitions)
+                    .filter(p -> InternalPartition.replicaIndices().allMatch(i -> p.getReplica(i) == null))
+                    .map(InternalPartition::getPartitionId)
+                    .collect(Collectors.toCollection(() -> new PartitionIdSet(partitions.length)));
+
             if (!partitionIds.isEmpty()) {
                 PartitionReplica[][] state = partitionStateManager.repartition(shutdownRequestedMembers, partitionIds);
                 if (state != null) {
                     logger.warning("Assigning new owners for " + partitionIds.size() + " LOST partitions!");
+
+                    int replicaUpdateCount = (int) partitionIds.stream()
+                            .flatMap(partitionId -> Arrays.stream(state[partitionId]).filter(Objects::nonNull)).count();
+                    MigrationStateImpl[] states = {new MigrationStateImpl(Clock.currentTimeMillis(), replicaUpdateCount, 0, 0L)};
+
+                    PartitionEventManager partitionEventManager = partitionService.getPartitionEventManager();
+                    partitionEventManager.sendMigrationProcessStartedEvent(states[0]);
+
                     partitionIds.intIterator().forEachRemaining((IntConsumer) partitionId -> {
                         InternalPartitionImpl partition = partitionStateManager.getPartitionImpl(partitionId);
                         PartitionReplica[] replicas = state[partitionId];
-
-                        assignLostPartitionOwner(partition, replicas[0]);
                         partition.setReplicas(replicas);
+
+                        InternalPartition.replicaIndices()
+                                .filter(i -> replicas[i] != null)
+                                .forEach(i -> {
+                                    MigrationInfo migration = new MigrationInfo(partitionId, null, replicas[i], -1, -1, -1, i)
+                                            .setStatus(MigrationStatus.SUCCESS);
+                                    states[0] = states[0].onComplete(0L);
```

Comment:
Reference of `MigrationStateImpl` reference 😃 `state **MigrationStateImpl`

Reply:
I can convert `InternalPartition.replicaIndices()` stream to a loop if you wish, but still `states[0]` array is needed because of outer `partitionIds.intIterator().forEachRemaining(...`.

Reply:
No need, it's not a big deal.